packages/react-dom/src/client/ReactDOMFloatClient.js

```diff
@@ -0,0 +1,814 @@
+        switch (loadingState.s) {
+          case 'l': {
+            resource.loaded = true;
+            resource.error = false;
+            break;
+          }
+          case 'e': {
+            resource.error = true;
+            break;
+          }
+          default: {
+            attachLoadListeners(existingEl, resource);
+          }
+        }
+      } else {
+        resource.loaded = true;
+      }
+    } else {
+      const instance = createResourceInstance(
+        'link',
+        resource.props,
+        ownerDocument,
+      );
+
+      attachLoadListeners(instance, resource);
+      insertStyleInstance(instance, precedence, ownerDocument);
+      resource.instance = instance;
+    }
+  }
+  resource.count++;
+  return resource.instance;
+}
+
+function attachLoadListeners(instance: Instance, resource: StyleResource) {
+  const listeners = {};
+  listeners.load = onResourceLoad.bind(
+    null,
+    instance,
+    resource,
+    listeners,
+    loadAndErrorEventListenerOptions,
+  );
+  listeners.error = onResourceError.bind(
+    null,
+    instance,
+    resource,
+    listeners,
+    loadAndErrorEventListenerOptions,
+  );
+
+  instance.addEventListener(
+    'load',
+    listeners.load,
+    loadAndErrorEventListenerOptions,
+  );
+  instance.addEventListener(
+    'error',
+    listeners.error,
+    loadAndErrorEventListenerOptions,
+  );
+}
+
+const loadAndErrorEventListenerOptions = {
+  passive: true,
+};
+
+function onResourceLoad(
+  instance: Instance,
+  resource: StyleResource,
+  listeners: {[string]: () => mixed},
+  listenerOptions: typeof loadAndErrorEventListenerOptions,
+) {
+  resource.loaded = true;
+  resource.error = false;
+  for (const event in listeners) {
+    instance.removeEventListener(event, listeners[event], listenerOptions);
+  }
+}
+
+function onResourceError(
+  instance: Instance,
+  resource: StyleResource,
+  listeners: {[string]: () => mixed},
+  listenerOptions: typeof loadAndErrorEventListenerOptions,
+) {
+  resource.loaded = false;
+  resource.error = true;
+  for (const event in listeners) {
+    instance.removeEventListener(event, listeners[event], listenerOptions);
+  }
+}
+
+function insertStyleInstance(
+  instance: Instance,
+  precedence: string,
+  ownerDocument: Document,
+): void {
+  const nodes = ownerDocument.querySelectorAll(
+    'link[rel="stylesheet"][data-rprec]',
+  );
+  const last = nodes.length ? nodes[nodes.length - 1] : null;
+  let prior = last;
+  for (let i = 0; i < nodes.length; i++) {
+    const node = nodes[i];
+    const nodePrecedence = node.dataset.rprec;
+    if (nodePrecedence === precedence) {
+      prior = node;
+    } else if (prior !== last) {
+      break;
+    }
+  }
+  if (prior && prior.parentNode) {
+    prior.parentNode.insertBefore(instance, prior.nextSibling);
+  } else {
+    const parent =
+      ownerDocument.head || ownerDocument.body || ownerDocument.documentElement;
```

Comment:
We've noticed similar patterns from extensions and other things but it ends up not really working as expected in the unexpected scenarios, or causing weird timing bugs when it's not inserted quickly enough.

I feel like we should just use `head` and error if it doesn't exist.

Reply:
Yeah just trying to be resilient but fine not supporting shenanigans that lead to no head